### ISSUE TITLE
OpenShift support

### ISSUE THREAD

**Issue by barkbay** [elastic/cloud-on-k8s#541]
Not really an issue but a way to track what I did to deploy the operator on Openshift : 

## Can't set finalizers without appropriate RBAC

If you try to deploy the operators on Openshift the following error is raised : 

```
2019-03-05T08:47:56.006Z        ERROR   kubebuilder.controller  Reconciler error        {"controller": "elasticsearch-controller", "request": "default/elasticsearch-sample", "error": "services \"elasticsearch-sample-es-discovery\" is forbidden: cannot set blockOwnerDeletion if an ownerReference refers to a resource you can't set finalizers on: no RBAC policy matched, <nil>", "errorCauses": [{"error": "services \"elasticsearch-sample-es-discovery\" is forbidden: cannot set blockOwnerDeletion if an ownerReference refers to a resource you can't set finalizers on: no RBAC policy matched, <nil>"}]}
```

It is because Openshift requires to explicitly set some permissions on the `xxx/finalizers`, resources.
To fix the problem add the finalizers sub-resource to the RBAC.
For example : 

```
- apiGroups:
  - elasticsearch.k8s.elastic.co
  resources:
  - elasticsearchclusters
  - elasticsearchclusters/status
  - elasticsearchclusters/finalizers  <---- here
  - clusterlicenses
  - trustrelationships
  - users
  - users/status
  verbs:
  - get
  - list
  - watch
  - create
  - update
  - patch
  - delete
```

## Users CRD

We should keep in mind that Openshift comes with its own `users` CRD, if a user want to manipulate Elasticsearch `users`  it should be done with the fully qualified name : 

```
(⎈ |default/192-168-99-100:8443/system:admin:default)➜  PR446 git:(e9496ac) ✗ k api-resources |grep users
users                                                elasticsearch.k8s.elastic.co            true         User
users                                                user.openshift.io                       false        User
```
See also https://github.com/elastic/k8s-operators/pull/446

## Default SCC does not allow privileged containers

SCC can be seen as the ancestor of [PSP](https://kubernetes.io/docs/concepts/policy/pod-security-policy/). Openshift clusters are deployed with a set of default SCCs.
The default SCC does not allow a pod/container to run in a privileged mode or to run with the `root` user (even worth, by default the containers are running with a random UID).
One (dirty) solution is to add the service account that need to be privileged in the `privileged` SCC : `oc adm policy add-scc-to-user privileged <sa>` 
A more proper way to do this would be to declare and deploy our own SCC (or PSP) to setup the right amount of privileges that are needed.

**Comment by mhofstetter** [comment]
Hi

I faced another problem installing the elastic operator in OpenShift after workaround the problems already listed.

## Elasticsearch Cert Initializer
The "CSR endpoint" from the init-container _cert-initializer_ isn't accessible by the elastic operator, if the operator and the elasticsearch instance aren't in the same namespace / project. Therefore the initialization is stuck in the init phase of the pod.

Link to cert initializer workflow illustration: https://github.com/elastic/cloud-on-k8s/tree/0.8/operators/cmd/cert-initializer

`
2019-06-04T06:57:03.765Z    INFO    certificate-initializer    No private key found on disk, will create one    {"reason": "open /mnt/elastic/private-key/node.key: no such file or directory"}                                                                            2019-06-04T06:57:03.765Z    INFO    certificate-initializer    Creating a private key on disk                                                                                                                                                                              2019-06-04T06:57:03.911Z    INFO    certificate-initializer    Generating a CSR from the private key                                                                                                                                                                        2019-06-04T06:57:03.913Z    INFO    certificate-initializer    Serving CSR over HTTP    {"port": 8001}                                                                                                                                                                      2019-06-04T06:57:03.914Z    INFO    certificate-initializer    Watching filesystem for cert update
`

This because by default OpenShift doesn't allow accessing pods directly between different namespaces / projects.

There are possibilities to configure OpenShift's SDN to allow the connection by "joining" the projects, but this involves the work of a clusteradmin for every new namespace / project where someone wants to setup an elasticsearch - which isn't ideal.

https://docs.openshift.com/container-platform/3.11/admin_guide/managing_networking.html

Maybe it would be a better approach to push the CSR from the POD (init container) to the elastic-operator (by providing a HTTP endpoint) instead of pulling it from the POD. This would only require that the elastic-operator is accessible from all other namespaces / projects - which can be achieved by defining the operator-namespace _elastic-system_ ONCE as _global_.

BTW: i didn't face this problem in minishift, only in OpenShift, since pods are accessible between namespaces in minishift by default. 

Elastic Operator: 0.8.0
OpenShift: 3.11

**Comment by barkbay** [comment]
https://github.com/elastic/cloud-on-k8s/pull/1083 should ensure that no code requiring root privileges is introduced in the operator and so we are still compatible with the `restricted` SCC of Openshift.

**Comment by barkbay** [comment]
A quick update on this one:

- Regarding Elasticsearch and Kibana I think we are good once #1136 is merged
- I'm working on some samples dedicated to Openshift, it could be useful to add some of them with some [Routes](https://docs.openshift.com/container-platform/3.11/architecture/networking/routes.html#overview) and with `setVmMaxMapCount: false`. I'm also wondering if a small user doc. would be a "nice to have", including some tips about the [Openshift SDN](https://docs.openshift.com/container-platform/3.11/architecture/networking/sdn.html#overview).
- Regarding the support of the APM server there is issue: the docker image of the APM server has not been designed to be run with a random UID, I will try to address that with a PR on the APM project

**Comment by barkbay** [comment]
PR for the support of the apm server is open here: https://github.com/elastic/apm-server/pull/2325

**Comment by barkbay** [comment]
An other issue has been in the beats project to clarify the situation around the Docker image: https://github.com/elastic/beats/issues/12686

**Comment by barkbay** [comment]
We still need some automated tests but I did some successful smoke tests with master today on Openshift 3.11
Let's close it for now and open a dedicated issue for automated tests.
